src/main/java/org/truffleruby/interop/InteropNodes.java

```diff
@@ -859,4 +883,85 @@ public Object metaObject(Object value) {
+        public boolean readableBit(int bits) {
+            return KeyInfo.isExisting(bits);
+        }
+
+    }
+
+    @CoreMethod(names = "readable_bit?", isModuleFunction = true, required = 1, lowerFixnum = 1)
+    public abstract static class HasReadableBitNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public boolean readableBit(int bits) {
+            return KeyInfo.isReadable(bits);
+        }
+
+    }
+
+    @CoreMethod(names = "writable_bit?", isModuleFunction = true, required = 1, lowerFixnum = 1)
+    public abstract static class HasWritableBitNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public boolean writableBit(int bits) {
+            return KeyInfo.isWritable(bits);
+        }
+
+    }
+
+    @CoreMethod(names = "invocable_bit?", isModuleFunction = true, required = 1, lowerFixnum = 1)
+    public abstract static class HasInvocableBitNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public boolean invocableBit(int bits) {
+            return KeyInfo.isInvocable(bits);
+        }
+
+    }
+
+    @CoreMethod(names = "internal_bit?", isModuleFunction = true, required = 1, lowerFixnum = 1)
+    public abstract static class HasInternalBitNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public boolean internalBit(int bits) {
+            return KeyInfo.isInternal(bits);
+        }
+
+    }
+
+    @CoreMethod(names = "key_info_flags_to_bits", isModuleFunction = true, required = 5)
+    public abstract static class KeyInfoFlagsToBitsNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization(guards = "existing")
+        public int keyInfoFlagsToBitsNode(boolean existing, boolean readable, boolean writable,
+                                          boolean invocable, boolean internal) {
+            return KeyInfo.newBuilder()
+                    .setReadable(readable)
+                    .setWritable(writable)
+                    .setInvocable(invocable)
+                    .setInternal(internal).build();
+        }
+
+        @Specialization(guards = {"!existing", "!readable", "!writable", "!invocable", "!internal"})
+        public int keyInfoFlagsToBitsNodeZero(boolean existing, boolean readable, boolean writable,
+                                              boolean invocable, boolean internal) {
+            return 0;
+        }
+
+        @Specialization(guards = {"!existing", "((readable || writable) || invocable) || internal"})
+        public int keyInfoFlagsToBitsNodeNotExisting(boolean existing, boolean readable, boolean writable,
+                                                     boolean invocable, boolean internal) {
+            return KeyInfo.newBuilder()
+                    .setReadable(readable)
+                    .setWritable(writable)
+                    .setInvocable(invocable)
+                    .setInternal(internal).build() & ~(KeyInfo.newBuilder().build());
```

Comment:
That builder API looks really sub-optimal. I think it's worth filing a Truffle bug.
There should just be a `setExisting()` method.

Reply:
Planning to.